src/checkpoint.rs

```diff
@@ -0,0 +1,141 @@
+use crate::errors::Result;
+use crate::mock::MockStorage;
+use crate::traits::{Api, Extern, ReadonlyStorage, Storage};
+
+pub struct Checkpoint<'a, S: Storage> {
+    /// a backing storage that is only modified upon commit
+    storage: &'a mut S,
+    /// these are local changes not flushed to backing storage
+    local_state: MockStorage,
+    /// this is a list of changes to be written to backing storage upon commit
+    rep_log: Vec<Op>,
+}
+
+enum Op {
+    Set { key: Vec<u8>, value: Vec<u8> },
+}
+
+impl<'a, S: Storage> Checkpoint<'a, S> {
+    pub fn new(storage: &'a mut S) -> Self {
+        Checkpoint {
+            storage,
+            local_state: MockStorage::new(),
+            rep_log: vec![],
```

Comment:
`local_state` and `rep_log` are only needed because we cannot loop through `local_state`'s keys? Otherwise we could get rid of the log and just write the latest entries of `local_state` into `storage`.

Reply:
Yeah, if we could loop through keys in a deterministic order, replog would not be needed.

This will likely change a bit when we extend the storage interface to allow iteration. Also if we add support for delete, not just set